### PR TITLE
Use memory-mapping in BasePreparedScriptStoreImpl::tryGetPreparedScript

### PR DESCRIPTION
**Pull request by JunielKatarn**
Uses `MemoryMappedBuffer` to retrieve prepared stored scripts.
This way, subsequent retievals of a prepared script / bytecode will avoid further disk access.

- Makes `Microsoft::JSI::MakeMemoryMappedBuffer` return `unique_ptr`.
- Updates `MemoryMappedBuffer` to drop Windows 7-required APIs.
- Adds minimal testing comparing stored and retrieved content sizes.

### Purpose
Required by Office as part of a series JavaScript store optimizations.

### Testing
- `Microsoft::JSI::Test::ScriptStoreIntegrationTest::RetrievePreparedScriptMemoryUsage`
  - Verifies the memory usage difference (working set) after calling `tryGetPreparedScript` is minimal and does not grow as a multiple of the retrieved file size (previous behavior).

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7757)

### CODE REVIEW

File: vnext/Shared/BaseScriptStoreImpl.cpp [hunk 1]
```diff
@@ -139,22 +144,7 @@ std::unique_ptr<const jsi::Buffer> LocalFileSimpleBufferStore::getBuffer(const s
...
-  std::streamsize size = file.tellg();
-  file.seekg(0, std::ios::beg);
-
-  auto buffer = std::make_unique<ByteArrayBuffer>(static_cast<size_t>(size));
-  if (!file.read(reinterpret_cast<char *>(buffer->data()), size)) {
-    return nullptr;
-  }
-
-  return buffer;
+  return Microsoft::JSI::MakeMemoryMappedBuffer(winrt::to_hstring(storeDirectory_ + bufferId).c_str());
```
Comment (hansenyy) [inline vnext/Shared/BaseScriptStoreImpl.cpp:147]: >winrt::to_hstring [](http://example.com/codeflow?start=48&length=17)

Does this method perform proper unicode conversion? I believe that storeDirectory and bufferId can both contain UTF-8 characters?

File: vnext/Desktop.UnitTests/ScriptStoreTests.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    constexpr size_t fileSize = 4 * 1024 * 1024;
+    constexpr size_t bufferSize = fileSize + 1; // +1 for trailing \0
+    char *buffer = static_cast<char *>(std::malloc(bufferSize * sizeof(char)));
+    int cursor = 0;
+    char id = 0;
+    while (cursor < fileSize) {
+      cursor += sprintf_s(buffer + cursor, bufferSize - cursor, "var v%019X = 'x';\n", id++);
+    }
+
+    auto stringBuffer = make_shared<StringBuffer>(buffer);
```
Comment (hansenyy) [inline vnext/Desktop.UnitTests/ScriptStoreTests.cpp:43]: >StringBuffer [](http://example.com/codeflow?start=36&length=12)

I am not sure what the StringBuffer constructor does, but please make sure that we are not doing a double free here.

File: vnext/Desktop.DLL/react-native-win32.x86.def [hunk 3]
```diff
@@ -25,7 +25,7 @@ EXPORTS
 ?InitializeTracing@react@facebook@@YGXPAUINativeTraceHandler@12@@Z
 ?Make@IWebSocketResource@React@Microsoft@@SG?AV?$shared_ptr@UIWebSocketResource@React@Microsoft@@@std@@$$QAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@5@@Z
 ?Make@JSBigAbiString@react@facebook@@SG?AV?$unique_ptr@$$CBUJSBigAbiString@react@facebook@@U?$default_delete@$$CBUJSBigAbiString@react@facebook@@@std@@@std@@$$QAV?$unique_ptr@U?$IAbiArray@D@AbiSafe@@UAbiObjectDeleter@2@@5@@Z
-?MakeMemoryMappedBuffer@JSI@Microsoft@@YG?AV?$shared_ptr@VBuffer@jsi@facebook@@@std@@QB_WI@Z
+?MakeMemoryMappedBuffer@JSI@Microsoft@@YG?AV?$unique_ptr@VBuffer@jsi@facebook@@U?$default_delete@VBuffer@jsi@facebook@@@std@@@std@@QB_WI@Z
```
Comment (NickGerleman) [inline vnext/Desktop.DLL/react-native-win32.x86.def:28]: Are we using this api in office?

File: vnext/Shared/BaseScriptStoreImpl.cpp (hunk 1 above)
Comment (NickGerleman) [reply vnext/Shared/BaseScriptStoreImpl.cpp:147]: to_hstring should do proper UTF-8 to UTF-16 conversion.

File: vnext/Desktop.UnitTests/ScriptStoreTests.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      Assert::Fail(L"Could not get temporary folder");
+    }
+
+    constexpr size_t fileSize = 4 * 1024 * 1024;
+    constexpr size_t bufferSize = fileSize + 1; // +1 for trailing \0
+    char *buffer = static_cast<char *>(std::malloc(bufferSize * sizeof(char)));
+    int cursor = 0;
+    char id = 0;
+    while (cursor < fileSize) {
+      cursor += sprintf_s(buffer + cursor, bufferSize - cursor, "var v%019X = 'x';\n", id++);
```
Comment (hansenyy) [inline vnext/Desktop.UnitTests/ScriptStoreTests.cpp:40]: >cursor += sprintf_s(buffer + cursor, bufferSize - cursor, "var v%019X = 'x';\n", id++); [](http://example.com/codeflow?start=6&length=87)

Does this line behave properly when sprintf_s fails (and returns -1) due to insufficient space remaining in the buffer?

Also, since this code does not need to be super performant, using std::stringstream might be more readable.

Lastly, do we even need to put valid JS in the buffer? From what I can tell, this test just test the storage and retrieval functionalities of BasePreparedScriptStoreImpl, so I think we can just use a random string for simplicity.

File: vnext/Desktop.UnitTests/ScriptStoreTests.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    char tempPath[MAX_PATH];
+    if (GetTempPathA(MAX_PATH, tempPath)) {
+      preparedScriptStore = make_unique<facebook::react::BasePreparedScriptStoreImpl>(tempPath);
+    } else {
+      Assert::Fail(L"Could not get temporary folder");
+    }
+
+    constexpr size_t fileSize = 4 * 1024 * 1024;
+    constexpr size_t bufferSize = fileSize + 1; // +1 for trailing \0
+    char *buffer = static_cast<char *>(std::malloc(bufferSize * sizeof(char)));
```
Comment (NickGerleman) [inline vnext/Desktop.UnitTests/ScriptStoreTests.cpp:36]: Why are we using a raw buffer instead of an raii type like vector? Note you can still manipulate a raw vector buffer.

**Review by hansenyy — CHANGES_REQUESTED** [review]
:clock1:

**Review by NickGerleman — APPROVED** [review]
Overall lgtm but left a couple of comments to look at.

File: vnext/Shared/BaseScriptStoreImpl.cpp (hunk 1 above)
Comment (hansenyy) [reply vnext/Shared/BaseScriptStoreImpl.cpp:147]: Hum I can't find any documentation that specifies that. @NickGerleman do you know where I can find it?

Also, as a separate issue, if to_hstring does proper unicode conversions, should we replace usages of the unicode conversion functions in Common with it?

**Review by hansenyy — APPROVED** [review]
:shipit:

**Comment by hansenyy** [comment]
> Overall lgtm but left a couple of comments to look at.

Same here.

File: vnext/Desktop.DLL/react-native-win32.x86.def (hunk 3 above)
Comment (JunielKatarn) [reply vnext/Desktop.DLL/react-native-win32.x86.def:28]: Not at all.
I checked before making this change instead of adding a new factory method.

File: vnext/Shared/BaseScriptStoreImpl.cpp (hunk 1 above)
Comment (vmoroz) [reply vnext/Shared/BaseScriptStoreImpl.cpp:147]: @hansenyy it does the proper conversion: you can see the source code for it in the base.h. Let's not change our code to use it yet.

**Review by vmoroz — APPROVED** [review]
:shipit:

File: vnext/Desktop.UnitTests/ScriptStoreTests.cpp (hunk 4 above)
Comment (JunielKatarn) [reply vnext/Desktop.UnitTests/ScriptStoreTests.cpp:40]: 1. It probably doesn't. I'll address this.
2. I'll follow up offline. Not sure how to combine `std::stringstream` with string formatting.
3. Not at all, but figured that out later on.

File: vnext/Desktop.UnitTests/ScriptStoreTests.cpp (hunk 5 above)
Comment (JunielKatarn) [reply vnext/Desktop.UnitTests/ScriptStoreTests.cpp:36]: Because of the string formatting, but @hansenyy makes a good point that we don't need proper JS for this test. Any type of content should work.

File: vnext/Desktop.DLL/react-native-win32.x86.def (hunk 3 above)
Comment (NickGerleman) [reply vnext/Desktop.DLL/react-native-win32.x86.def:28]: Can we remove remove it from the ABI then, or do we still need it for the  test project?

File: vnext/Shared/BaseScriptStoreImpl.cpp (hunk 1 above)
Comment (JunielKatarn) [reply vnext/Shared/BaseScriptStoreImpl.cpp:147]: Yes, we should. I have done that in a few modules.
Regarding the docs, I infer the function does proper UTF conversion from the following:

[`winrt::to_string`](https://docs.microsoft.com/en-us/uwp/cpp-ref-for-winrt/to-string) uses [WideCharToMultiByte](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-widechartomultibyte) which seems to ensure UTF conversion.\
It's therefore very likely that [`winrt::to_hstring`](https://docs.microsoft.com/en-us/uwp/cpp-ref-for-winrt/to-hstring) uses [MultiByteToWideChar](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-multibytetowidechar), which should ensure proper UTF-8 to UTF-16 conversion.

This is also suggested by @kennykerr [here](https://kennykerr.ca/2018/02/28/cppwinrt-strings/).

I'll ping the C++/WinRT before merging.

File: vnext/Desktop.DLL/react-native-win32.x86.def (hunk 3 above)
Comment (JunielKatarn) [reply vnext/Desktop.DLL/react-native-win32.x86.def:28]: Ideally, we should test it as an integration test (react-native-win32.dll), but other types used by the test are not in the ABI and we don't want to add them just for testing purposes.

TL;DR
We can remove it at least for now.

File: vnext/Shared/BaseScriptStoreImpl.cpp (hunk 1 above)
Comment (kennykerr) [reply vnext/Shared/BaseScriptStoreImpl.cpp:147]: Yep, those helpers properly handle utf conversions back and forth.

File: vnext/Desktop.UnitTests/ScriptStoreTests.cpp (hunk 2 above)
Comment (JunielKatarn) [reply vnext/Desktop.UnitTests/ScriptStoreTests.cpp:43]: It does not. I made sure.
That said, I'll switch away from the `char*` based on your previous suggestion.

**Comment by JunielKatarn** [comment]
@asklar See `Purpose` in description.
